harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,437 @@
+
+        if pynvml:
+            self._init_pynvml()
+        else:
+            logging.warning(f"pynvml module not found. GPU metrics will not be collected.")
+
+    def _init_pynvml(self) -> None:
+        """
+        Call the NVML library methods we'll be using to validate that they are accessible.
+        Sometimes NVML initializes successfully but individual device methods will fail.
+        """
+        try:
+            pynvml.nvmlInit()
+            num_gpus = pynvml.nvmlDeviceGetCount()
+            for i in range(num_gpus):
+                handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+                uuid = pynvml.nvmlDeviceGetUUID(handle)
+                pynvml.nvmlDeviceGetMemoryInfo(handle)
+                pynvml.nvmlDeviceGetUtilizationRates(handle)
+                self._pynvml_device_handles[uuid] = handle
+        except pynvml.NVMLError as ne:
+            logging.info(f"Error accessing NVML {ne}. GPU metrics will not be collected.")
+        except Exception as e:
+            logging.error(f"Error initializing pynvml: {e}")
+
+    def collect(self) -> Dict[str, Any]:
+        metrics = {}
+
+        for uuid, handle in self._pynvml_device_handles.items():
+            gpu_util = pynvml.nvmlDeviceGetUtilizationRates(handle).gpu
+            free_memory = pynvml.nvmlDeviceGetMemoryInfo(handle).free
+            metrics.update(
+                {
+                    uuid: {
+                        "gpu_util": gpu_util,
+                        "gpu_free_memory": free_memory,
+                    }
+                }
+            )
+        return metrics
+
+
+class _Metric:
+    def __init__(
+        self,
+        group: str,
+        metrics: Dict[str, Any],
+        timestamp: datetime,
+    ):
+        self.group = group
+        self.metrics = metrics
+        self.timestamp = timestamp
+
+
+class _Collector(threading.Thread):
+    def __init__(
+        self,
+        metrics_queue: queue.Queue,
+        collection_interval: int = 1,
+        max_collection_s: int = 60 * 30,
+    ):
+        self._collection_interval = collection_interval
+        self._should_exit = False
+        self._last_flush_ts = None
+        self._metrics_queue = metrics_queue
+        self._max_collection_s = max_collection_s
+
+        self._metric_collectors = [
+            _GPU(),
+            _CPU(),
+            _Memory(),
+            _Disk(),
+            _Network(),
+        ]
+        self._should_exit = threading.Event()
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        collection_stop_ts = time.time() + self._max_collection_s
+        while not self._should_exit.is_set():
+            if time.time() > collection_stop_ts:
+                # Max collection time reached. Notify queue consumers and exit.
+                self._metrics_queue.put(None)
+                return
+
+            self._collect_metrics()
+            self._should_exit.wait(self._collection_interval)
```

Comment:
does this need a treatment similar to https://github.com/determined-ai/determined/pull/8964/files#diff-fe44d33f23b5de232f042f28c2a78ca4b0b869cb1244c84189edf2d0867ba209R400
e.g. if the metric collection from nvml or others is not super fast, it'd be better to sleep for `1 - time to collect metrics`.

Reply:
very good point. changed this logic to account for time taken to collect metrics.